File: lecture_31/p4.cpp
#include<iostream>
#include<queue>
#include<vector>
using namespace std;

// https://practice.geeksforgeeks.org/problems/top-view-of-binary-tree/1

void topView(struct Node *root)
{
    unordered_map<int,int> map1;
    int mi=0;
    int ma=0;
    queue<pair<Node*,int>> que;
    if(root!=nullptr)
        que.push(pair<Node*,int>(root,0));
    while(que.size()>0)
    {
        int size=que.size();
        while(size-->0)
        {
            pair<Node*,int> rp=que.front();
            que.pop();
            mi=min(mi,rp.second);
            ma=max(ma,rp.second);
            if(map1.count(rp.second)==0)
                map1[rp.second]=rp.first->data;
            if(rp.first->left!=nullptr)
                que.push(pair<Node*,int> (rp.first->left,rp.second-1));
            if(rp.first->right!=nullptr)
                que.push(pair<Node*,int> (rp.first->right,rp.second+1));
        }
    }
    for(int i=mi;i<=ma;i++)
        cout<<map1[i]<<" ";
}

int main(int args,char** argv)
{
    return 0;
}

File: lecture_30/p4.java
public class p4 {
    int ans=Integer.MIN_VALUE;
    public int solve(TreeNode node)
    {
        if(node==null)
            return 0;
        int ls=solve(node.left);
        int rs=solve(node.right);
        ans=Math.max(ans,node.val);
        ans=Math.max(ans,node.val+ls);
        ans=Math.max(ans,node.val+rs);
        ans=Math.max(ans,node.val+ls+rs);
        return Math.max(node.val,Math.max(node.val+ls,node.val+rs));
    }
    public int maxPathSum(TreeNode root) {
        solve(root);
        return ans;
    }
    
    public static void main(String[] args) {
        
    }
}